src/box/memtx_tx.c

```diff
@@ -2017,16 +2017,42 @@ memtx_tx_handle_gap_write(struct space *space, struct memtx_story *story,
 				   index->def->cmp_def->part_count;
 		bool is_eq = item->type == ITER_EQ || item->type == ITER_REQ;
 		bool is_e = item->type == ITER_LE || item->type == ITER_GE;
+
+		/*
+		 * The key+iterator pair matches to the new tuple and can match
+		 * to its predecessors, but it can also be matched with anything
+		 * else between this tuple and the successor (exclusively).
+		 */
 		bool need_split = item->key == NULL ||
 				  (dir * cmp > 0 && !is_eq) ||
 				  (!is_full_key && cmp == 0 && (is_e || is_eq));
+
+		/*
+		 * The key+iterator pair can only match the new tuple, and its
+		 * predecessors, but not the successor's predecessors anymore.
+		 */
 		bool need_move = !need_split &&
 				 ((dir < 0 && cmp > 0) ||
 				  (cmp > 0 && item->type == ITER_EQ) ||
-				  (cmp == 0 && ((dir < 0 && is_full_key) ||
+				  (cmp == 0 && ((is_eq && is_full_key) ||
```

Comment:
Doesn't it belong to the previous commit?

Reply:
No, these changes are introduced because the `count_fast` implementation (introduced in this commit) does not use the `memtx_tx_track_point` for readability reasons.